### PR TITLE
Lint

### PR DESCRIPTION
**Pull request by nick-drozd**
This PR unpins the pylint version and fixes all resulting lint errors but one. That last error is fixed in https://github.com/hyperledger/sawtooth-core/pull/612, which needs to be merged before this one.

Notes:

1) This PR doesn't extend `run_lint` coverage, so there are still files not being linted.

2) The current version of pylint sometimes triggers a recursion error and spits out a big ugly recursion error stack trace. This makes reading the output very annoying, but it doesn't cause `run_lint` to fail.

### CODE REVIEW

File: bin/protogen [hunk 1]
```diff
@@ -34,7 +34,8 @@ JOIN = os.path.join
 TOP_DIR = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
 
 
-def main(args=sys.argv[1:]):
+def main(args=None):
+    args = sys.argv[1:] if args is None else args
```
Comment (vaporos) [inline bin/protogen]: Avoid using conditional expressions as replacements for simplistic flow control statements.  This is better:

```
if args is None:
    args = sys.argv[1:]
```

This is much clearer and technically correct (doesn't re-assign args when it not None).  It is also fewer tokens to read total and fewer tokens-per-line.

File: consensus/poet/core/sawtooth_poet/poet_consensus/poet_config_view.py [hunk 2]
```diff
@@ -139,6 +139,7 @@ def enclave_module_name(self):
                     name='sawtooth.poet.enclave_module_name',
                     value_type=str,
                     default_value=PoetConfigView._ENCLAVE_MODULE_NAME_,
+                    # pylint: disable=len-as-condition
```
Comment (vaporos) [inline consensus/poet/core/sawtooth_poet/poet_consensus/poet_config_view.py]: Could cast to bool here instead of the pylint comment.

File: validator/sawtooth_validator/state/client_handlers.py [hunk 3]
```diff
@@ -82,7 +82,7 @@ def __init__(self, request_proto, response_proto, response_type,
         self._block_store = block_store
         self._batch_cache = batch_cache
 
-    def handle(self, identity, message_content):
+    def handle(self, connection_id, message_content):
```
Comment (aludvik) [inline validator/sawtooth_validator/state/client_handlers.py]: Need to update the docstring below.

File: rest_api/sawtooth_rest_api/config.py [hunk 4]
```diff
@@ -124,4 +124,4 @@ def to_dict(self):
         ])
 
     def to_toml_string(self):
-        return toml.dumps(self.to_dict()).strip().split('\n')
+        return str(toml.dumps(self.to_dict())).strip().split('\n')
```
Comment (aludvik) [inline rest_api/sawtooth_rest_api/config.py]: Can you include some notes in your commit message about why this was necessary?

File: validator/sawtooth_validator/journal/block_store.py [hunk 5]
```diff
@@ -130,10 +130,9 @@ def get_predecessor_iter(self, starting_block=None):
         Returns:
             An iterator
         """
-        if not starting_block:
-            return _BlockPredecessorIterator(self, start_block=self.chain_head)
-        else:
-            return _BlockPredecessorIterator(self, start_block=starting_block)
+        start_block = starting_block if starting_block else self.chain_head
```
Comment (aludvik) [inline validator/sawtooth_validator/journal/block_store.py]: IMO, introducing another variable like this adds an unnecessary level of indirection. I prefer something like the following:
```
if not starting_block:
    return _BlockPredecessorIterator(self, start_block=self.chain_head)

return _BlockPredecessorIterator(self, start_block=starting_block)
```

File: validator/sawtooth_validator/journal/chain.py [hunk 6]
```diff
@@ -105,10 +105,12 @@ def __init__(self,
...
     def _get_previous_block_root_state_hash(self, blkw):
         if blkw.previous_block_id == NULL_BLOCK_IDENTIFIER:
-            return INIT_ROOT_KEY
+            state_hash = INIT_ROOT_KEY
         else:
             prev_blkw = self._block_cache[blkw.previous_block_id]
-            return prev_blkw.state_root_hash
+            state_hash = prev_blkw.state_root_hash
+
+        return state_hash
```
Comment (aludvik) [inline validator/sawtooth_validator/journal/chain.py]: IMO this is better:
```
if blkw.previous_block_id == NULL_BLOCK_IDENTIFIER:
    return INIT_ROOT_KEY

return self._block_cache[blkw.previous_block_id].state_root_hash
```

ditto for the remainder of this commit.

File: validator/sawtooth_validator/journal/consensus/dev_mode/dev_mode_consensus.py [hunk 7]
```diff
@@ -140,6 +140,7 @@ def finalize_block(self, block_header):
 class BlockVerifier(BlockVerifierInterface):
     """DevMode BlockVerifier implementation
     """
+    # pylint: disable=useless-super-delegation
```
Comment (aludvik) [inline validator/sawtooth_validator/journal/consensus/dev_mode/dev_mode_consensus.py]: Rather than disable these, can we update the Interface classes to have concrete constructors and then just delete the `__init__`  methods here?

File: validator/sawtooth_validator/database/lmdb_nolock_database.py [hunk 8]
```diff
@@ -54,6 +54,7 @@ def __init__(self, filename, flag):
                                       create=create,
                                       lock=True)
 
+    # pylint: disable=no-value-for-parameter
```
Comment (aludvik) [inline validator/sawtooth_validator/database/lmdb_nolock_database.py]: Can this be resolved instead of disabling it?

File: validator/sawtooth_validator/journal/block_store.py (hunk 5 above)
Comment (jjason) [inline validator/sawtooth_validator/journal/block_store.py]: What about:
```
if not starting_block:
    starting_block = self.chain_head

return _BlockPredecessorIterator(self, start_block=starting_block)
```

Comment (jjason) [inline validator/sawtooth_validator/journal/block_store.py]: I'll retract my comment....after thinking about it more, I would go with @aludvik recommendation so we don't give lint the opportunity to complain about re-assigning the `starting_block` parameter in a future version.

File: consensus/poet/core/sawtooth_poet/poet_consensus/poet_config_view.py [hunk 9]
```diff
@@ -139,7 +139,8 @@ def enclave_module_name(self):
                     name='sawtooth.poet.enclave_module_name',
                     value_type=str,
                     default_value=PoetConfigView._ENCLAVE_MODULE_NAME_,
-                    validate_function=lambda value: len(value) > 0)
+                    # function should return true if value is nonempty
+                    validate_function=lambda value: value)
```
Comment (nick-drozd) [inline consensus/poet/core/sawtooth_poet/poet_consensus/poet_config_view.py]: Ha, you thought it would be that easy?
```
************* Module sawtooth_poet.poet_consensus.poet_config_view
W:143,38: Lambda may not be necessary (unnecessary-lambda)
```

File: validator/sawtooth_validator/database/lmdb_nolock_database.py (hunk 8 above)
Comment (nick-drozd) [inline validator/sawtooth_validator/database/lmdb_nolock_database.py]: Not in any obvious way. The thing works as far as I can tell, so the error might be spurious. In any case, I've alerted @jsmitchell, who can take further action if need be.

File: validator/sawtooth_validator/journal/consensus/dev_mode/dev_mode_consensus.py (hunk 7 above)
Comment (nick-drozd) [inline validator/sawtooth_validator/journal/consensus/dev_mode/dev_mode_consensus.py]: All of the obvious ways of changing this result in further pylint errors (pylint seems to disagree with this design in general). Nontrivial changes are outside the scope of this PR. I've alerted @jjason.

File: rest_api/sawtooth_rest_api/config.py (hunk 4 above)
Comment (nick-drozd) [inline rest_api/sawtooth_rest_api/config.py]: Added

File: validator/sawtooth_validator/state/client_handlers.py (hunk 3 above)
Comment (nick-drozd) [inline validator/sawtooth_validator/state/client_handlers.py]: Updated

File: validator/sawtooth_validator/journal/consensus/dev_mode/dev_mode_consensus.py (hunk 7 above)
Comment (nick-drozd) [inline validator/sawtooth_validator/journal/consensus/dev_mode/dev_mode_consensus.py]: (Actually, @cianx made it, but my general point remains.)

File: cli/sawtooth_cli/exceptions.py [hunk 10]
```diff
@@ -15,5 +15,4 @@
 
 
 class CliException(Exception):
-    def __init__(self, msg):
-        super(CliException, self).__init__(msg)
+    pass
```
Comment (nick-drozd) [inline cli/sawtooth_cli/exceptions.py]: I reworded the relevant commit to reflect this.

File: validator/sawtooth_validator/database/lmdb_nolock_database.py (hunk 8 above)
Comment (jsmitchell) [inline validator/sawtooth_validator/database/lmdb_nolock_database.py]: I guess you could try `return txn.stat(self._lmdb)['entries']` to see if it works.
